### PR TITLE
Add some clarifications to README.md in the database schema directory

### PR DESCRIPTION
**Pull request by rubo77**
Please someone explain, what the two folders 16 and 54 are about and why the dump is used for later, and why it is important.

I plan to add some information, how to add the foreign key CONSTRAINT like visualized in https://github.com/matrix-org/synapse/issues/934 like

    ALTER TABLE event_forward_extremities ADD CONSTRAINT room_id_fk FOREIGN KEY (room_id) REFERENCES rooms (room_id);
    ALTER TABLE room_account_data         ADD CONSTRAINT room_id_fk FOREIGN KEY (room_id) REFERENCES rooms (room_id);
    ALTER TABLE room_account_data         ADD CONSTRAINT user_id_fk FOREIGN KEY (user_id) REFERENCES users (name);
    ALTER TABLE room_memberships          ADD CONSTRAINT room_id_fk FOREIGN KEY (room_id) REFERENCES rooms (room_id);
    ALTER TABLE room_memberships          ADD CONSTRAINT user_id_fk FOREIGN KEY (user_id) REFERENCES users (name);
    ALTER TABLE room_stats_state          ADD CONSTRAINT room_id_fk FOREIGN KEY (room_id) REFERENCES rooms (room_id);
    ALTER TABLE state_groups              ADD CONSTRAINT room_id_fk FOREIGN KEY (room_id) REFERENCES rooms (room_id);

Flooring keys would be a great way to maintain data integrity, especially as a safeguard against orphaned records.

Maybe we can cleanup script that removes all orphaned entries and then adds all foreign keys to the PostgreSQL Database, or is there already such a script out in the wild somewhere?

### CODE REVIEW

File: synapse/storage/data_stores/main/schema/full_schemas/README.md [hunk 1]
```diff
@@ -1,13 +1,23 @@
 # Building full schema dumps
 
-These schemas need to be made from a database that has had all background updates run.
+These schemas need to be made from a database that has had all background
+updates run. The schemas are used at ...
```
Comment (richvdh) [inline synapse/storage/data_stores/main/schema/full_schemas/README.md]: at ... ?

File: synapse/storage/data_stores/main/schema/full_schemas/README.md [hunk 2]
```diff
@@ -1,13 +1,23 @@
...
 
 To do so, use `scripts-dev/make_full_schema.sh`. This will produce
-`full.sql.postgres ` and `full.sql.sqlite` files.
+`full.sql.postgres ` and `full.sql.sqlite` files. It generates SQL files for
+creating a brand new Synapse DB with the latest schema, on both SQLite3 and
+Postgres. It does so by having Synapse generate an up-to-date SQLite DB, then
```
Comment (richvdh) [inline synapse/storage/data_stores/main/schema/full_schemas/README.md]: do we need this here as well as in the comments in the script itself? Having it in two places feels like a recipe for it getting out of date.

File: synapse/storage/data_stores/main/schema/full_schemas/README.md [hunk 3]
```diff
@@ -1,13 +1,23 @@
...
 
 Ensure postgres is installed and your user has the ability to run bash commands
 such as `createdb`.
 
 ```
 ./scripts-dev/make_full_schema.sh -p postgres_username -o output_dir/
 ```
+
+The exported database structure has to be cleared of obsolete tables:
+Delete the CREATE statements for "sqlite_stat1", "schema_version", "applied_schema_deltas", and "applied_module_schemas".
```
Comment (richvdh) [inline synapse/storage/data_stores/main/schema/full_schemas/README.md]: delete from where? I don't think this is correct.

File: synapse/storage/data_stores/main/schema/full_schemas/README.md [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
+
+There are two folders `16` and `54` which contain ...
```
Comment (richvdh) [inline synapse/storage/data_stores/main/schema/full_schemas/README.md]: which contain... ?

I assume you're referring to the directories in https://github.com/matrix-org/synapse/tree/master/synapse/storage/data_stores/main/schema/full_schemas.

54 is the most recent full-schema snapshot. 16 is a snapshot from 2015 or so, for historical reference,

**Comment by richvdh** [comment]
> I plan to add some information, how to add the foreign key CONSTRAINT 

please don't do this. It will lead to people having different database schemas to what we test and develop against, which will be a support nightmare.

I'm not totally against having foreign-key constraints in the database, but if we do it, we should do it for everyone via delta files, not by manual hackery. In any case, please come and discuss why we should do this in #synapse-dev first, once people are back from the holiday period.

File: synapse/storage/data_stores/main/schema/full_schemas/README.md (hunk 1 above)
Comment (rubo77) [reply synapse/storage/data_stores/main/schema/full_schemas/README.md]: the question is, why is this schema here in the repo? Is it used somewhere? Or just for information if someone is looking into the database structure?

File: synapse/storage/data_stores/main/schema/full_schemas/README.md (hunk 2 above)
Comment (rubo77) [reply synapse/storage/data_stores/main/schema/full_schemas/README.md]: I think, this readme is the right place for such information, maybe we can delete the info in the script top, but this especially doesent seem like some text that gets outdated

File: synapse/storage/data_stores/main/schema/full_schemas/README.md (hunk 1 above)
Comment (richvdh) [reply synapse/storage/data_stores/main/schema/full_schemas/README.md]: how do you think synapse sets up its database?

File: synapse/storage/data_stores/main/schema/full_schemas/README.md (hunk 2 above)
Comment (richvdh) [reply synapse/storage/data_stores/main/schema/full_schemas/README.md]: I'm afraid I disagree. How it works is irrelevant here.

File: synapse/storage/data_stores/main/schema/full_schemas/README.md [hunk 5]
```diff
@@ -1,13 +1,20 @@
...
 Ensure postgres is installed and your user has the ability to run bash commands
-such as `createdb`.
+such as `createdb`, then call
+
+    ./scripts-dev/make_full_schema.sh -p postgres_username -o output_dir/
 
-```
-./scripts-dev/make_full_schema.sh -p postgres_username -o output_dir/
-```
+There are two folders with full-schema snapshots `16` is a snapshot from 2015,
```
Comment (richvdh) [inline synapse/storage/data_stores/main/schema/full_schemas/README.md]:
```suggestion
There are currently two folders with full-schema snapshots. `16` is a snapshot from 2015,
```

File: synapse/storage/data_stores/main/schema/full_schemas/README.md [hunk 6]
```diff
@@ -1,13 +1,20 @@
...
+for historical reference, and one folder with most recent database-schema.
```
Comment (richvdh) [inline synapse/storage/data_stores/main/schema/full_schemas/README.md]:
```suggestion
for historical reference. `54` is currently the most recent full schema snapshot.
```

File: synapse/storage/data_stores/main/schema/full_schemas/README.md [hunk 7]
```diff
@@ -1,13 +1,20 @@
-# Building full schema dumps
+# Synapse Database Schemas
 
-These schemas need to be made from a database that has had all background updates run.
+These schemas are used to set up a brand new Synapse DB on both SQLite3 and
```
Comment (richvdh) [inline synapse/storage/data_stores/main/schema/full_schemas/README.md]:
```suggestion
These schemas are used as a basis to create brand new Synapse databases, on both SQLite3 and
```

File: synapse/storage/data_stores/main/schema/full_schemas/README.md [hunk 8]
```diff
@@ -1,13 +1,20 @@
...
+Postgres
```
Comment (richvdh) [inline synapse/storage/data_stores/main/schema/full_schemas/README.md]:
```suggestion
Postgres.
```

File: changelog.d/6615.misc [hunk 9]
```diff
@@ -0,0 +1 @@
+Explain more in database structure README.md
```
Comment (richvdh) [inline changelog.d/6615.misc]: this isn't even a sentence

```suggestion
Add some clarifications to `README.md` in the database schema directory.
```

File: synapse/storage/data_stores/main/schema/full_schemas/README.md [hunk 10]
```diff
@@ -1,13 +1,20 @@
...
+To do so, use `scripts-dev/make_full_schema.sh`. This will produce new
+`full.sql.postgres ` and `full.sql.sqlite` files. 
 
 Ensure postgres is installed and your user has the ability to run bash commands
-such as `createdb`.
+such as `createdb`, then call
+
+    ./scripts-dev/make_full_schema.sh -p postgres_username -o output_dir/
 
-```
```
Comment (richvdh) [inline synapse/storage/data_stores/main/schema/full_schemas/README.md]: triple-backtick works fine here, but :man_shrugging:

Comment (rubo77) [reply synapse/storage/data_stores/main/schema/full_schemas/README.md]: it saves a line of code and is better readeable if not parsed

File: synapse/storage/data_stores/main/schema/full_schemas/README.md (hunk 6 above)
Comment (rubo77) [reply synapse/storage/data_stores/main/schema/full_schemas/README.md]: I'll change it to "The other contains the most recent full
schema snapshot." so we wouldn't have to change the Readme every time, we update the schema

Comment (richvdh) [reply synapse/storage/data_stores/main/schema/full_schemas/README.md]: but this will be wrong next time we create a full schema... :man_shrugging:

Comment (rubo77) [reply synapse/storage/data_stores/main/schema/full_schemas/README.md]: Why is this wron the next time? "the other" is always the latest, isn't it?

Comment (richvdh) [reply synapse/storage/data_stores/main/schema/full_schemas/README.md]: yes, but then there will be three folders.

Comment (richvdh) [reply synapse/storage/data_stores/main/schema/full_schemas/README.md]: the point is: we're going to have to update this paragraph next time we create a full schema anyway, so we may as well be clear which folders we are talking about.

**Pull request merged**